drivers/eeprom/eeprom_emulator.c

```diff
@@ -0,0 +1,744 @@
+
+	addr_jmp = address & (sizeof(buf) - 1);
+	if (addr_jmp) { /* unaligned start */
+		size_t plen = sizeof(buf) - addr_jmp;
+
+		rc = eeprom_emu_word_read(dev, address - addr_jmp, buf);
+		if (rc) {
+			goto end;
+		}
+
+		if (len < plen) {
+			plen = len;
+		}
+
+		if (write_op) {
+			memcpy(buf + addr_jmp, data8, plen);
+			rc = eeprom_emu_word_write(dev, address - addr_jmp,
+						   buf);
+			if (rc) {
+				goto end;
+			}
+
+		} else {
+			memcpy(data8, buf + addr_jmp, plen);
+		}
+
+		len -= plen;
+		data8 += plen;
+		address += plen;
+	}
+
+	while (len >= sizeof(buf)) { /* aligned part */
+		if (write_op) {
+			rc = eeprom_emu_word_write(dev, address, data8);
+		} else {
+			rc = eeprom_emu_word_read(dev, address, buf);
+			memcpy(data8, buf, sizeof(buf));
+		}
+
+		if (rc) {
+			goto end;
+		}
+
+		data8 += sizeof(buf);
+		address += sizeof(buf);
+		len -= sizeof(buf);
+	}
+
+	if (len) {
+		/* unaligned end */
+		rc = eeprom_emu_word_read(dev, address, buf);
+		if (rc) {
+			goto end;
+		}
+
+		if (write_op) {
+			memcpy(buf, data8, len);
+			rc = eeprom_emu_word_write(dev, address, buf);
+		} else {
+			memcpy(data8, buf, len);
+		}
+
+	}
+
+end:
+	k_mutex_unlock(&dev_data->lock);
+	if (write_op) {
+		(void)flash_write_protection_set(dev_config->flash_dev, true);
+	}
+
+	return rc;
+}
+
+static int eeprom_emu_read(const struct device *dev, off_t address, void *data,
+			   size_t len)
+{
+	return eeprom_emu_rw(dev, address, data, len, false);
+}
+
+static int eeprom_emu_write(const struct device *dev, off_t address,
+			    const void *data, size_t len)
+{
+	return eeprom_emu_rw(dev, address, (void *)data, len, true);
+}
+
+static size_t eeprom_emu_size(const struct device *dev)
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+
+	return dev_config->size;
+}
+
+static int eeprom_emu_init(const struct device *dev)
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+	struct eeprom_emu_data *dev_data = DEV_DATA(dev);
+	off_t offset;
+	uint8_t buf[dev_config->flash_cbs];
+	int rc;
+
+	k_mutex_init(&dev_data->lock);
+	if (!device_is_ready(dev_config->flash_dev)) {
+		__ASSERT(0, "Could not get flash device binding");
+		return -ENODEV;
+	}
+
+	/* Find the page offset */
+	dev_data->page_offset = 0U;
+	dev_data->write_offset = dev_config->page_size - sizeof(buf);
+	while (dev_data->page_offset < dev_config->flash_size) {
+		offset = dev_data->page_offset + dev_data->write_offset;
+		rc = eeprom_emu_flash_read(dev, offset, buf, sizeof(buf));
+		if (rc) {
+			return rc;
+		}
+
+		if (!eeprom_emu_is_word_used(dev, buf)) {
+			break;
+		}
+
+		dev_data->page_offset += dev_config->page_size;
+	}
+
+	if (dev_data->page_offset == dev_config->flash_size) {
+		__ASSERT(0, "All pages are invalid, is this a EEPROM area?");
+		return -EINVAL;
+	}
+
+	dev_data->write_offset = dev_config->size;
+
+	/* No need to update the write offset for readonly */
+	if (dev_config->readonly) {
+		return 0;
```

Comment:
Wont we miss filling in the optional RAM buffer on read-only devices?

Reply:
Correct, I have corrected this by removing this early return. This will also allow a emulated eeprom that has changes written to flash to make a transition from rw to readonly.